src/neo/SmartContract/ContractTask.cs

```diff
@@ -0,0 +1,42 @@
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace Neo.SmartContract
+{
+    class ContractTask : INotifyCompletion
+    {
+        private Action continuation;
+
+        public bool IsCompleted => false;
+
+        public ContractTask GetAwaiter() => this;
+
+        public void GetResult() { }
+
+        public void OnCompleted(Action continuation)
+        {
+            Interlocked.CompareExchange(ref this.continuation, continuation, null);
+        }
+
+        public void RunContinuation()
+        {
+            continuation();
+        }
+    }
+
+    class ContractTask<T> : ContractTask
```

Comment:
We may need to add an AsyncMethodBuilder type in order to compose ContractTask<T> types. Docs on building "task-like types" is here: https://github.com/dotnet/roslyn/blob/master/docs/features/task-types.md

Reply:
I don’t understand, why do we need `AsyncMethodBuilder` now that it can work normally?

Reply:
AsyncMethodBuilder is used to build async methods that return ContractTask. For example, ContractManagement.OnDeploy awaits the call to CallFromNativeContract. C# has provides AsyncMethodBuilder implementations for void, Task/Task<T> and ValueTask/ValueTask<T> methods.